ECOM-7388 Program details page redesign

MatthewPiatetsky opened the pull request:
Pretty much all the files are new. The idea is there are old files and new files with same name and a _2017 suffix. When we're ready to remove the old code, we just delete all the old files, and remove the _2017 from the new files.
Most of the changes are in underscore templates and css. 
Most of the views with _2017 are nearly the same as the old ones, just renamed so we can clean up easily.

[Demo](http://18.189.48.145:8000/dashboard/programs/0ffff5d6-0177-4690-9a48-aa2fecf94610/)
email: `mpiatetsky@edx.org`
password: edx

Tasks
- [x] a11y requirements review
- [x] finalize responsive design
- [x] final design review
- [x] analytics
- [x] browser testing
- - [x] Chrome
- - [x] Safari
- - [x] Mobile Safari
- - [x] Firefox
- - [ ] Edge
- - [x] IE 11
- [ ] test button functionality
- [x] tests
- [x] verify edge cases where data is missing
 (missing dates, empty course sections)

AlasdairSwan commented on lms/static/images/programs/micromasters-program-details.svg:
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1524.05 150.17"><defs><style>.cls-1{fill:#005585;}</style></defs><title>micromasters-icon</title><path class="cls-1" d="M31.57,65.82a3,3,0,0,0,.72.09,3,3,0,0,0,2.91-2.28,74.73,74.73,0,0,1,14-28.45A143.33,143.33,0,0,0,63.68,39a75.1,75.1,0,0,0-8.46,23.65,3,3,0,1,0,5.91,1.06A68,68,0,0,1,70,40.24a216.58,216.58,0,0,0 ...
```
I would remove the id and data-name attributes, and the ```<defs>``` (including the styles). Could you add a more meaningful class than ```cls-1``` and add a default fill colour to Sass.

AlasdairSwan commented on lms/static/images/programs/professional-certificate-program-details.svg:
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2116.96 151"><defs><style>.cls-1{fill:#9a1f60;}</style></defs><title>professional-certificate-icon-SR</title><path class="cls-1" d="M219,42A42,42,0,0,0,153.33,7.33,3,3,0,0,0,152,7H3a3,3,0,0,0-3,3V125a3,3,0,0,0,3,3H146v20a3,3,0,0,0,5.12,2.12l25.41-25.41,12,11.69c3.55,3.44,7.11,6.88,9.79,9.46C203.71,151,203.71,151 ...
```
I would remove the id and data-name attributes, and the ```<defs>``` (including the styles). Could you add a more meaningful class than ```cls-1``` and add a default fill colour to Sass.

AlasdairSwan commented on lms/static/images/programs/xseries-program-details.svg:
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1198.33 151"><defs><style>.cls-1{fill:#424242;}</style></defs><title>xseries-icon-SR</title><path class="cls-1" d="M219,42A42,42,0,0,0,153.33,7.33,3,3,0,0,0,152,7H3a3,3,0,0,0-3,3V125a3,3,0,0,0,3,3H146v20a3,3,0,0,0,5.12,2.12l25.41-25.41,12,11.69c3.55,3.44,7.11,6.88,9.79,9.46C203.71,151,203.71,151,205,151a3,3,0,0, ...
```
I would remove the id and data-name attributes, and the ```<defs>``` (including the styles). Could you add a more meaningful class than ```cls-1``` and add a default fill colour to Sass.

AlasdairSwan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,30 @@
+                    var verified_seat, currency;
+                    if ('seats' in run && run['seats'].length) {
+                        verified_seat = _.filter(run['seats'], function(seat){
+                            if (seat['type'] === 'verified' || seat['type'] === 'professional' || seat['type'] === 'credit'){
```
Dot notation is preferred (e.g. ```seat.type```).

AlasdairSwan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,30 @@
+                                return seat;
+                            }
+                        })[0]
+                        currency = verified_seat['currency'];
```
Dot notation please.

AlasdairSwan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,30 @@
+                        })[0]
+                        currency = verified_seat['currency'];
+                        if (currency === 'USD'){
+                            return '$' + verified_seat['price'];
```
Dot notation please.

AlasdairSwan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,30 @@
+                        if (currency === 'USD'){
+                            return '$' + verified_seat['price'];
+                        } else {
+                            return verified_seat['price'] + " " + currency;
```
Nitpick: single quotes should be used instead of double wherever possible in keeping with the coding style of the repo.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/collection_list_view.js:
```diff
@@ -47,6 +47,10 @@
                             this.$el.before(HtmlUtils.ensureHtml(this.getTitleHtml()).toString());
                         }
 
+                        count = this.$el.parent().find('.count');
```
Views should not really be updating HTML outside themselves. The View that contains ```.count``` should be listening for a model change or something and update it itself.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/course_card_view_2017.js:
```diff
@@ -0,0 +1,84 @@
+
+                 render: function() {
+                     var data = $.extend(this.model.toJSON(), {enrolled: this.context.enrolled || ''});
+                     var filledTemplate = this.tpl(data);
```
There should only be 1 var declaration at the top of the function, comma separating the variable names.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/course_enroll_view_2017.js:
```diff
@@ -0,0 +1,111 @@
+                     this.render();
+                     if (this.urlModel) {
+                         this.trackSelectionUrl = this.urlModel.get('track_selection_url');
+                     }
```
It seems like render should be called after all of the other code in the init function.

MatthewPiatetsky replied:
i just copied this from the previous file, hard to tell if there was any reason to put it after. should i put this in a postRender function?

AlasdairSwan replied:
Yeah, I think it makes sense to move it.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/course_enroll_view_2017.js:
```diff
@@ -0,0 +1,111 @@
+                 handleEnroll: function() {
+                    // Enrollment click event handled here
+                     if (!this.model.get('course_run_key')) {
+                         this.$('.select-error').css('visibility', 'visible');
```
I would prefer to use a CSS class that's added or removed instead of setting the element's style directly.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/program_header_view_2017.js:
```diff
@@ -0,0 +1,76 @@
+                 },
+
+                 getLogo: function() {
+                    var logo,
```
Should it default to false in case a 4th program type is returned in error.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+
+.program-details {
+    .window-wrap {
+        background-color: rgb(255,255,255);
```
Please use a Sass variable.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+.program-details-wrapper {
+
+    .program-details-header {
+        background-color: rgb(252,252,252);
```
Please use a Sass variable.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        }
+
+        .micromasters {
+            fill: #065784;
```
Please use a Sass variable.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        }
+
+        .xseries {
+            fill: #065784;
```
Please use a Sass variable.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        }
+
+        .professional.certificate {
+            fill: #065784;
```
Please use a Sass variable.

rlucioni commented on lms/djangoapps/learner_dashboard/views.py:
```diff
@@ -66,6 +67,7 @@ def program_details(request, program_uuid):
 
     context = {
         'program_data': program_data,
+        'courses_progress': meter.progress(programs=meter.programs, count_only=False)[0],
```
There's going to be a lot duplicated between `program_data` and `courses_progress`. The bulk of `program_data` is data about the courses in the program, and it looks like that same data is going to end up grouped in `course_progress`. What can we do about that?

I'd update the tests of this view so that they verify the data is present on the page. Also, if these end up staying separate, a more descriptive name might be nice here. `course_groups`, for example.

MatthewPiatetsky replied:
well i think the program data is primarily used for the program data and this is just the course data.  
Assuming we didn't need the course data we could theoretically remove it from the program_data object but that doesn't seem necessary to me

rlucioni replied:
@MatthewPiatetsky the course data is the majority of the program data. If nothing changes here, we'll send that data to clients twice for no reason. We shouldn't increase the size of the response more than we have to.

MatthewPiatetsky replied:
are you suggesting we just delete the course data from the object if it is in fact unused?

rlucioni replied:
@MatthewPiatetsky not necessarily, I'm just asking you to find a way to avoid increasing the size of the response unnecessarily. That could mean popping the course data from the program data dict if it's not going to be messy.

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -211,11 +214,12 @@ def completed_course_runs(self):
         """
         course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
-        return [
+        completed_course_runs = [
```
Is this necessary?

rlucioni commented on openedx/core/djangoapps/programs/tests/test_utils.py:
```diff
@@ -53,7 +53,7 @@ def _create_enrollments(self, *course_run_ids):
     def _assert_progress(self, meter, *progresses):
         """Variadic helper used to verify progress calculations."""
-        self.assertEqual(meter.progress, list(progresses))
+        self.assertEqual(meter.progress(), list(progresses))
```
A test that covers the case where `count_only` is False would be good.

MatthewPiatetsky replied:
i can't figure out how to run this test in paver, the command in the docs is failing for me for some reason

MatthewPiatetsky replied:
added

AlasdairSwan commented on lms/static/sass/partials/base/_variables.scss:
```diff
@@ -224,6 +224,11 @@ $success-color: rgb(0, 155, 0) !default;
+// logo colors
+$micromasters-logo-blue: #005585;
+$xseries-logo-gray: #424242;
+$professional-certificate-logo-red: #9a1f60;
```
Nitpick: I'm not sure it matters that they are logos or what the name of the color is (e.g. if next month they wanted to make the xseries color orange) so I would consider renaming them ``$micromasters-color```, etc.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        @media(min-width: $bp-screen-md) {
+            margin-left: 30px;
+            margin-right: 80px;
+        }
```
I would move this down a few lines so it's after the initial margin declarations. Also, an alternative way of writing this is;
```
margin: {
    left: 30px;
    right: 80px;
}
```

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        padding-top: 40px;
+        padding-bottom: 35px;
+        display: flex;
+        color: black;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+    }
+
+    .course-list > div:nth-of-type(even) {
+        background-color: #f9fafc;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+        .divider {
+            margin-left: 0;
+            margin-bottom: 20px;
+            background-color: #BBC7D2;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            font-family: "Open Sans";
+            font-weight: 600;
+            font-size: 1.3em;
+            color: #373737;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            font-weight: 600;
+            font-size: 1.3em;
+            color: #373737;
+            margin-bottom: 0 0 5px 0;
```
This should either be ```margin-bottom: 5px``` or ```margin: 0 0 5px 0;```

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            }
+
+            button {
+                background-color: #3C5E7E;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+                background-color: #3C5E7E;
+                height: 37px;
+                width: 128px;
+                border-radius: 0;
```
Is it not 0 by default?

AlasdairSwan replied:
Cool

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            font-family: "Open Sans";
+            font-weight: bold;
+            font-size: 0.9375em;
+            color: #2c3e50;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+
+            .course-title { 
+                font-size: 1em;
+                color: #0976b5;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            }
+
+            .run-period {
+                color: #292929;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+                    font-family: "Open Sans";
+                    font-weight: bold;
+                    font-size: 0.9375em;
+                    color: #2c3e50;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+                }
+
+                .price {
+                    color: #006d01;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+
+        .certificate-status {
+            .fa-check-circle {
+                color: #006d01;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+                font-family: "Open Sans";
+                font-weight: bold;
+                font-size: 0.9375em;
+                color: #2c3e50;
```
Variables should be used for colours.

AlasdairSwan commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,375 @@
+            }
+
+            .certificate-status-msg {
+                color: #292929;
```
Variables should be used for colours.

mikedikan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,32 @@
+                getCertificatePriceString: function(run) {
+                    var verifiedSeat, currency;
+                    if ('seats' in run && run.seats.length) {
+                        // eslint-disable-next-line consistent-return
```
can you leave a comment explaining why we ignore the linter here?

MatthewPiatetsky replied:
I am using filter and not returning all seat types (which means the function doesn't always return.)  
Not sure, maybe i can return nothing to make the filter ignore it or something like that

mikedikan replied:
that sounds reasonable

mikedikan commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,32 @@
+                            }
+                        })[0];
+                        currency = verifiedSeat.currency;
+                        if (currency === 'USD') {
```
USD is the only currency where the sign precedes the value?

MatthewPiatetsky replied:
for the other currencies we just have the name. 

It would be weird to put something like USD10. This is basically only handling USD and adding a relatively sane fallback behavior for other currencies until we know how we want to deal with them

mikedikan replied:
👍

mikedikan commented on lms/templates/discovery/course_card_2017.underscore:
```diff
@@ -0,0 +1,24 @@
+            </h2>
+            <div class="course-date" aria-hidden="true">
+                <%- interpolate(
+                      gettext("Starts: %(start_date)s"),
```
what is the 's' after the start_date variable?

MatthewPiatetsky replied:
@rlucioni do you know?

AlasdairSwan replied:
@mikedikan @MatthewPiatetsky it's the wrapper for vars that interpolate expects ```Starts: %(my_var)s```

mikedikan commented:
@MatthewPiatetsky can you update the demo link so it uses your IP?

MatthewPiatetsky commented:
@AlasdairSwan i think i've either commented on or addressed most of your comments.  
Also, do you know where I would look to add tests for this page?

AlasdairSwan commented:
@MatthewPiatetsky here are the test for the current program dashboard https://github.com/edx/edx-platform/tree/26201161813a15b565ffdccd91c1ea5ee9a01b0e/lms/static/js/spec/learner_dashboard

AlasdairSwan commented on lms/static/js/learner_dashboard/views/course_card_view_2017.js:
```diff
@@ -0,0 +1,84 @@
+                 },
+
+                 render: function() {
+                     var data = $.extend(this.model.toJSON(), {enrolled: this.context.enrolled || ''}),
```
Would it make more sense to default the value of enrolled to ```false```?

MatthewPiatetsky replied:
at the moment i was passing the text to display to the template. would it be better if it was a boolean that that the template looked at with an if block?

AlasdairSwan replied:
That's fine, I was assuming it was a boolean and not text.

AlasdairSwan commented on lms/static/js/learner_dashboard/views/program_details_view_2017.js:
```diff
@@ -34,31 +33,64 @@
                          model: new Backbone.Model(this.options)
                      });
+
+                     new CollectionListView({
```
Shouldn't this be wrapped in an ```if (this.remainingCourseCollection.length > 0) {``` ?

AlasdairSwan commented on lms/templates/learner_dashboard/certificate-status-2017.underscore:
```diff
@@ -0,0 +1,5 @@
+    <% // safe-lint: disable=underscore-not-escaped %>
+    <span class="certificate-icon green-icon" aria-hidden="true"><%= certificateSvg %></span>
+    <span class="card-msg"><%- gettext('Congratulations! You have earned a certificate for this course.') %></span>
+</div>
```
Why 2 versions of this file?

AlasdairSwan commented on lms/templates/learner_dashboard/course_enroll_2017.underscore:
```diff
@@ -0,0 +1,51 @@
+<% if (is_enrolled) { %>
+<% } else { %>
```
Shouldn't this just be ```<% if (!is_enrolled) { %>```?

MatthewPiatetsky commented on lms/static/js/learner_dashboard/models/course_card_model.js:
```diff
@@ -115,12 +115,32 @@
                     if (courseRun) {
-                        if (courseRun.advertised_start !== undefined && courseRun.advertised_start !== 'None') {
+
+                        if (!([undefined, 'None', null].indexOf(courseRun.advertised_start) >= 0)) {
```
this line is checking that courseRun.advertised_start is not undefined, 'None' or null  
do you have suggestion for making this more clear?

AlasdairSwan replied:
```if (!_.isUndefined(course_run.advertised_start))```

MatthewPiatetsky replied:
that works for the undefined case but not the other two I think?

AlasdairSwan replied:
Yeah, you're right. The 'None' value is a weird one. I think it would be clearer what's going on in the code if you created a function called something like ```isDefined``` that does the check i.e.
```
isDefined: function(str) {
    return [undefined, 'None', null].indexOf(str) >= 0;
}
```

AlasdairSwan replied:
@MatthewPiatetsky I would still prefer this check to be its own function to make the code easier to read.

MatthewPiatetsky replied:
thanks changed

MatthewPiatetsky commented on lms/static/js/learner_dashboard/views/program_header_view_2017.js:
```diff
@@ -0,0 +1,76 @@
+        'edx-ui-toolkit/js/utils/html-utils',
+        'text!../../../templates/learner_dashboard/program_header_view_2017.underscore',
+        'picturefill',
+        'text!/static/images/programs/micromasters-program-details.svg',
```
@AlasdairSwan if i comment out these images my program_details_view_spec.js test work  
otherwise the tests error on a 404 on these. what is the correct way to include them in the spec file?

AlasdairSwan replied:
That's weird, I didn't think we have to do anything special because the same ```text!``` plugin is being used for underscore templates. Maybe the karma.conf isn't aware of the folder the images are in?

MatthewPiatetsky commented on lms/static/js/spec/learner_dashboard/course_enroll_view_spec_2017.js:
```diff
@@ -0,0 +1,293 @@
+define([
```
we already had this file, so this is almost a carbon copy with some small tweaks for the new code

MatthewPiatetsky commented:
@AlasdairSwan @rlucioni bump

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -136,19 +138,31 @@ def progress(self, programs=None, count_only=True):
                 progress, and unstarted courses instead of serialized representations
                 of the courses.
 
+            show_expired_runs_as_remaining (bool): Whether in progress runs whose upgrade
```
We don't want this method's output to change across the list and detail views. The user's view of progress should be consistent across those pages. This should just be how the method behaves, not an option.

We're also at risk of introducing debt here. It's not safe to assume that just because the upgrade deadline is past, completion of the course run will no longer count for completion of the course. Each program has a type, and those types define a [set of seat types](https://github.com/edx/course-discovery/blob/master/course_discovery/apps/course_metadata/models.py#L649) that qualify for completion of the program. Given that we have support for that, we shouldn't be hardcoding assumptions that all programs require verified seats like you're doing here.

At the very least, this should behave like the ProgramProgressMeter's `completed_programs` property. Read [this](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/programs/utils.py#L207-L213).

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -136,19 +138,31 @@ def progress(self, programs=None, count_only=True):
+                    if show_expired_runs_as_remaining:
+                        enrolled_run = [run for run in course['course_runs'] if run['is_enrolled']][0]
+                        enrolled_run_seats = [seat['upgrade_deadline'] for seat in enrolled_run['seats']]
+                        run_expired = any((x is not None) and (parse(x) < now) for x in enrolled_run_seats)
```
Please use intelligible variable names.

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -136,19 +138,31 @@ def progress(self, programs=None, count_only=True):
             list of dict, each containing information about a user's progress
                 towards completing a program.
         """
+        now = datetime.datetime.now(tzlocal())
```
Is passing `tzlocal()` required? Django should take care of timezone detection for you.

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -136,19 +138,31 @@ def progress(self, programs=None, count_only=True):
                     completed.append(course)
                 elif self._is_course_in_progress(course):
+                    if show_expired_runs_as_remaining:
+                        enrolled_run = [run for run in course['course_runs'] if run['is_enrolled']][0]
```
This is only going to work for an "extended_program." In other words, this will only work if you provide extended programs when initializing the class. Expecting people to know that when using this method worries me.

Furthermore, how are these course runs ordered? My sense is that it's not safe to just pick the first one. Consider the case where the user is enrolled in two runs of the same course. This may pick the older enrollment and miss the newer one.

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -136,19 +138,31 @@ def progress(self, programs=None, count_only=True):
                 elif self._is_course_in_progress(course):
+                    if show_expired_runs_as_remaining:
+                        enrolled_run = [run for run in course['course_runs'] if run['is_enrolled']][0]
+                        enrolled_run_seats = [seat['upgrade_deadline'] for seat in enrolled_run['seats']]
```
This variable name can be better; this isn't a list of seats.

rlucioni commented on lms/static/js/learner_dashboard/views/course_enroll_view_2017.js:
```diff
@@ -50,33 +49,26 @@
 
                  enrollSuccess: function() {
                      var courseRunKey = this.model.get('course_run_key');
+                     window.analytics.track('Program Details Enrollment', {
```
We use a specific format for events that we should stick to unless there's a good reason not to. Refer to the events in the edx.org GA account for an example.

Also: I don't think the `courseRun` property you're including here is going to be surfaced in GA. It may not be worth including.

MatthewPiatetsky replied:
ok, there are a bunch in different formats like 'Completed Purchase', but i'm assuming you mean the edx.bi format
i'll go with edx.bi.user.program-details.enrollment

MatthewPiatetsky commented:
@rlucioni you asked to bump you again

AlasdairSwan commented on lms/static/images/programs/professional-certificate-program-details.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2116.96 151"><title>Professional Certificate Logo</title><path d="M219,42A42,42,0,0,0,153.33,7.33,3,3,0,0,0,152,7H3a3,3,0,0,0-3,3V125a3,3,0,0,0,3,3H146v20a3,3,0,0,0,5.12,2.12l25.41-25.41,12,11.69c3.55,3.44,7.11,6.88,9.79,9.46C203.71,151,203.71,151,205,151a3,3,0,0,0,3-3V71a3,3,0,0,0-.07-.62A41.84,41.84,0,0,0,219,42Zm-6,0A36,36,0,1,1,177,6,36,36, ...
```
The svgs are still using ```cls-1``` as their class, please update to something more descriptive.

MatthewPiatetsky replied:
thanks fixed, before when i addressed your comment i didn't realize there were multiple instances of these classes in this file

AlasdairSwan commented on lms/static/js/learner_dashboard/views/course_card_view_2017.js:
```diff
@@ -0,0 +1,83 @@
+                 },
+
+                 render: function() {
+                     var data = $.extend(this.model.toJSON(), {enrolled: this.context.enrolled || ''});
```
Nitpick: the code would be easier to read if spread across a few lines, e.g.
```
var data = $.extend(this.model.toJSON(), {
    enrolled: this.context.enrolled || ''
});
```

MatthewPiatetsky replied:
changed

AlasdairSwan commented on lms/static/js/learner_dashboard/views/program_header_view_2017.js:
```diff
@@ -0,0 +1,76 @@
+                 },
+
+                 postRender: function() {
+                    // To resolve a bug in IE with picturefill reevaluate images
```
Are we still loading multiple versions of an image for responsive? If not we can remove this (and any other picturefill references).

MatthewPiatetsky replied:
good call, i don't think this is necessary anymore

AlasdairSwan commented on package.json:
```diff
@@ -20,8 +20,8 @@
-    "eslint-config-edx": "^2.0.0",
-    "eslint-config-edx-es5": "^2.0.0",
+    "eslint": "^2.13.1",
+    "eslint-config-edx": "^1.2.1",
```
Why these changes?

MatthewPiatetsky replied:
accidental commit, removed

rlucioni commented on lms/static/js/learner_dashboard/views/program_details_view_2017.js:
```diff
@@ -34,32 +33,71 @@
 
                  render: function() {
-                     HtmlUtils.setHtml(this.$el, this.tpl());
+                     var completedCount = this.completedCourseCollection.length,
```
@AlasdairSwan you may want to re-use this for the sidebar ring.

rlucioni commented on lms/static/js/learner_dashboard/views/program_header_view_2017.js:
```diff
@@ -0,0 +1,57 @@
+                     var logo = false,
+                         type = this.model.get('programData').type;
+
+                     if (type === 'MicroMasters') {
```
Is it easy to do this without hardcoding the names of the program types? We've been over why we try to avoid this before, so I won't elaborate.

MatthewPiatetsky replied:
not sure about this, asked Alasdair

AlasdairSwan replied:
It's going to be difficult because of the way we are passing in the SVG logos. I suppose we could have a logoCollection, and we could do ```logoCollection.findWhere({type: type})``` so at least we are moving the hard-coded program types into the logo Collection.

MatthewPiatetsky replied:
@rlucioni do you think that would be better? we would still have to hardcode the program types

rlucioni replied:
@MatthewPiatetsky no, it doesn't really solve the problem, just moves it elsewhere. Maybe it's not worth worrying about here, but I still think we should be aware of it.

AlasdairSwan replied:
The only other solution I can think of would be for the API to return the SVG.

rlucioni replied:
@AlasdairSwan good idea. The discovery API is currently capable of returning a link to an image for each program type inline with the serialized programs, and I could see us returning a link to an SVG served by discovery/S3. Right now there are PNGs logos for the program types stored in the discovery service.

Let's table this for now, but I (and @clintonb) like the idea of using the logos from the discovery service where possible instead of duplicating them in other services.

MatthewPiatetsky replied:
the complicating issue there is we may have a legitimate need to have different logos in different places (which we currently do). so it wouldn't be as simple as just storing a single logo. The logos would have to be tied somehow to where they would be used

rlucioni replied:
@MatthewPiatetsky yeah, that's a good point. Though I don't think it prevents discovery from providing an API that can get you the right image for the right context.

rlucioni commented on lms/static/js/learner_dashboard/views/program_header_view_2017.js:
```diff
@@ -0,0 +1,57 @@
+                     } else if (type === 'Professional Certificate') {
+                         logo = ProfessionalCertificateLogo;
+                     }
+                     return logo;
```
What if an open source user is trying to run this? The default is just not to show a logo at all?

MatthewPiatetsky replied:
that was my assumption. we don't have a default logo.

rlucioni replied:
Okay, just making sure I understand the intention. Seems fine to me.

rlucioni commented on lms/static/sass/views/_program-details.scss:
```diff
@@ -76,3 +76,425 @@
+        }
+
+    }
+}
```
Nitpick: missing newline.

MatthewPiatetsky replied:
fixed

rlucioni commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -158,7 +159,7 @@ def progress(self, programs=None, count_only=True):
                 'uuid': program['uuid'],
                 'completed': len(completed) if count_only else completed,
                 'in_progress': len(in_progress) if count_only else in_progress,
-                'not_started': len(not_started) if count_only else not_started,
```
Nitpick: please leave this trailing comma, it's good practice. It leaves the next person to touch this dictionary with a cleaner diff.

MatthewPiatetsky replied:
fixed
